### PR TITLE
More exact picking for canvas editor

### PR DESCRIPTION
**Pull request by ghost**
Currently all object picking in the canvas editor happens using rectangles.

This PR changes this to more exact checks (if an object is actually hit by a click), which should allow for better selection of objects:

https://www.dropbox.com/s/46jfqqyqkoyagir/demo-pick-canvas-2d.mp4?dl=0

### CODE REVIEW

File: scene/2d/canvas_item.h [hunk 1]
```diff
@@ -229,6 +229,7 @@ class CanvasItem : public Node {
 	// Used to resize/move/select the node
 	virtual void _edit_set_rect(const Rect2 &p_rect){};
 	virtual Rect2 _edit_get_rect() const { return Rect2(-32, -32, 64, 64); };
+	virtual bool _edit_is_clicked(const Point2 &p_point, double p_tolerance) const { return true; }
```
Comment (groud) [inline scene/2d/canvas_item.h]: `_edit_is_clicked` is a little bit confusing I think, as it makes me think the function is going to return whether the object is actually clicked or not. (thus depending on the actual state of the mouse).
Maybe we could go for something a little bit like "_edit_is_selected_on_click", so it is more obvious what the function does. (It is more understandable in this case that the object is not clicked yet)

File: scene/2d/navigation_polygon.cpp [hunk 2]
```diff
@@ -35,9 +35,50 @@
 
 #include "thirdparty/misc/triangulator.h"
 
+Rect2 NavigationPolygon::_edit_get_rect() const {
```
Comment (groud) [inline scene/2d/navigation_polygon.cpp]: Same here.

File: scene/2d/line_2d.cpp [hunk 3]
```diff
@@ -48,6 +48,32 @@ Line2D::Line2D() :
 	_round_precision = 8;
 }
 
+Rect2 Line2D::_edit_get_rect() const {
```
Comment (groud) [inline scene/2d/line_2d.cpp]: I'm pretty sure you cannot do that without modifying the _edit_set_rect function. I would leave it as is for now.

File: scene/2d/navigation_polygon.cpp [hunk 4]
```diff
@@ -311,6 +358,16 @@ bool NavigationPolygonInstance::is_enabled() const {
 
 /////////////////////////////
 
+Rect2 NavigationPolygonInstance::_edit_get_rect() const {
```
Comment (groud) [inline scene/2d/navigation_polygon.cpp]: Same here.

File: scene/2d/path_2d.cpp [hunk 5]
```diff
@@ -32,6 +32,51 @@
 #include "engine.h"
 #include "scene/scene_string_names.h"
 
+#ifdef TOOLS_ENABLED
+#include "editor/editor_scale.h"
+#endif
+
+Rect2 Path2D::_edit_get_rect() const {
```
Comment (groud) [inline scene/2d/path_2d.cpp]: Same here

File: scene/resources/capsule_shape_2d.cpp [hunk 6]
```diff
@@ -32,6 +32,25 @@
...
+		if (i == 6 || i == 18)
+			points.push_back(Vector2(Math::sin(i * Math_PI * 2 / 24.0), Math::cos(i * Math_PI * 2 / 24.0)) * get_radius() - ofs);
+	}
+
+	return points;
+}
+
+bool CapsuleShape2D::_edit_is_clicked(const Point2 &p_point, double p_tolerance) const {
+
+	return Geometry::is_point_in_polygon(p_point, _get_points());
```
Comment (groud) [inline scene/resources/capsule_shape_2d.cpp]: Wut ? Calculating whether a point is in a capsule or not does not require you to create a polygon...
Just check whether it is in the "body rectangle" then in the two extermity circles.

File: scene/resources/concave_polygon_shape_2d.cpp [hunk 7]
```diff
@@ -32,6 +32,23 @@
 #include "servers/physics_2d_server.h"
 #include "servers/visual_server.h"
 
+bool ConcavePolygonShape2D::_edit_is_clicked(const Point2 &p_point, double p_tolerance) const {
+
```
Comment (groud) [inline scene/resources/concave_polygon_shape_2d.cpp]: Is it checking for the inside of the polygon or only it's outline ?

**Review by groud — CHANGES_REQUESTED** [review]
A very good PR. Thanks ! :)
I wanted to implement a similar approach.

I think for now I'd prefer this PR not to override the _edit_get_rect in any case, this require more work. Indeed we have to decide whether the rectangle, when set, should modify the points or the scale in specific cases. For now I'm pretty sure it breaks the resizing mechanism.

File: scene/resources/concave_polygon_shape_2d.cpp (hunk 7 above)
Comment (poke1024) [reply scene/resources/concave_polygon_shape_2d.cpp]: It only checks the outline as drawn in `ConcavePolygonShape2D::draw()`; it's consistent, as the inside is not drawn.

File: editor/plugins/canvas_item_editor_plugin.cpp [hunk 8]
```diff
@@ -612,6 +612,7 @@ void CanvasItemEditor::_find_canvas_items_at_pos(const Point2 &p_pos, Node *p_no
 	if (Object::cast_to<Viewport>(p_node))
 		return;
 
+	const real_t grab_distance = EDITOR_DEF("editors/poly_editor/point_grab_radius", 8);
```
[the remark this answers is not in the document]
Comment (poke1024) [reply editor/plugins/canvas_item_editor_plugin.cpp]: Note that this is not a new option, but the standard handle grab radius that exists for all handles. If a user changes that to something larger than the default to allow more tolerance for clicking on handles, it makes much sense that the selection of lines and similar items should operate with the same tolerance.

File: scene/2d/line_2d.cpp (hunk 3 above)
Comment (poke1024) [reply scene/2d/line_2d.cpp]: Leaving it as it is means it's not selectable. The standard transform-based `Node2D::_edit_set_rect()` works fine in all these cases.

File: scene/resources/capsule_shape_2d.cpp (hunk 6 above)
Comment (poke1024) [reply scene/resources/capsule_shape_2d.cpp]: I know, but why spend time optimizing this and introducing just another possibility of mismatch between drawing and selection code.

File: scene/2d/canvas_item.h (hunk 1 above)
Comment (poke1024) [reply scene/2d/canvas_item.h]: Changed.

**Comment by poke1024** [comment]
Concerning the new overrides of `_edit_get_rect`; the behavior with the new items (`Line2D`, etc.) is exactly the same as it is with with `Polygon2D` in the current master, so I don't see any problem.

File: editor/plugins/canvas_item_editor_plugin.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (groud) [reply editor/plugins/canvas_item_editor_plugin.cpp]: Well, if it's already there I guess it's ok. 
We will remove it when we do a review of all editor's options I guess.

File: scene/resources/capsule_shape_2d.cpp (hunk 6 above)
Comment (groud) [reply scene/resources/capsule_shape_2d.cpp]: Well, you're building a polygon to check for 2 circles and a square, this is not optimizing. 
Honestly, if the drawing and the selection code mismatch a little bit it is not important. The code here is overly complex for little gain. Just keep thing simple, in this case maintainability is probably more important than accuracy. :)

**Comment by groud** [comment]
Ok then. I'm still not sure it's the best thing to modify the scale when resizing a polygon, line, etc... But if it is already like that ok. If it becomes a problem later on we can change it anyway. :)

**Pull request merged**